// eslint-plugin-sf-plugin/dist/rules/command-example.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.commandExamples = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const utils_1 = require("@typescript-eslint/utils");
const commands_1 = require("../shared/commands");
const createRule = utils_1.ESLintUtils.RuleCreator((name) => `https://example.com/rule/${name}`);
exports.commandExamples = createRule({
    name: 'command-example',
    meta: {
        docs: {
            description: 'Ensure commands have a summary, description, and examples',
            recommended: 'error',
        },
        messages: {
            example: 'Commands should have an examples property',
        },
        type: 'problem',
        schema: [],
    },
    defaultOptions: [],
    create(context) {
        return (0, commands_1.isInCommandDirectory)(context)
            ? {
                ClassDeclaration(node) {
                    // verify it extends SfCommand
                    if ((0, commands_1.extendsSfCommand)(node) && node.id) {
                        if (!node.body.body.some((member) => (0, commands_1.getClassPropertyIdentifierName)(member) === 'examples')) {
                            context.report({
                                node: node.id,
                                messageId: 'example',
                            });
                        }
                    }
                },
            }
            : {};
    },
});
