// eslint-plugin-sf-plugin/dist/rules/migration/no-number-flags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noNumberFlags = void 0;
/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
const utils_1 = require("@typescript-eslint/utils");
const commands_1 = require("../../shared/commands");
const flags_1 = require("../../shared/flags");
exports.noNumberFlags = utils_1.ESLintUtils.RuleCreator.withoutDocs({
    meta: {
        docs: {
            description: 'Change number flag to integer',
            recommended: 'error',
        },
        messages: {
            message: 'number flags are not available on sfCommand.  Use integer instead',
        },
        type: 'problem',
        schema: [],
        fixable: 'code',
    },
    defaultOptions: [],
    create(context) {
        return (0, commands_1.isInCommandDirectory)(context)
            ? {
                Property(node) {
                    if ((0, flags_1.isFlag)(node) && (0, commands_1.ancestorsContainsSfCommand)(context.getAncestors())) {
                        const toReplace = (0, flags_1.getCalleePropertyByName)(node, 'number');
                        if (toReplace) {
                            context.report({
                                node: toReplace,
                                messageId: 'message',
                                fix: (fixer) => fixer.replaceText(toReplace, 'integer'),
                            });
                        }
                    }
                },
            }
            : {};
    },
});
